### PR TITLE
Document safe rules in declarativeNetRequest

### PR DESCRIPTION
**Pull request by oliverdunk**
Documents safe rules which were implemented following this proposal: https://docs.google.com/document/d/1srkkCJkl4X2KOOUwnpDd-kvm8AY70eZBAYI4m836bvQ/edit?usp=sharing

As a note, some of the links here are dependent on `nodoc` being removed in the Chromium source to make constants visible in our documentation.

Fixes https://github.com/GoogleChromeLabs/Extension-Docs/issues/218

### CODE REVIEW

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md [hunk 1]
```diff
@@ -223,7 +229,18 @@ The number of rules available after that depends on how many rules are enabled b
 
 #### Dynamic and session rules {: #dynamic-session-rules }
 
-The limits applied to dynamic and session rules are simpler than static rules. The total number of both cannot exceed 5000. This is called the [`MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES`](#property-MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES).
+An extension can have up to 5000 session rules. This is exposed as the
+[`MAX_NUMBER_OF_SESSION_RULES`](#property-MAX_NUMBER_OF_SESSION_RULES).
+
+For dynamic rules, up to 30,000 rules are permitted, exposed as the
+[`MAX_NUMBER_OF_DYNAMIC_RULES`](#property-MAX_NUMBER_OF_DYNAMIC_RULES). Included in this set of
```
Comment (sebastianbenz) [inline site/en/docs/extensions/reference/declarativeNetRequest/index.md:236]: Do I understand this right, that the 5000 count towards the 30000 limit?

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md [hunk 2]
```diff
@@ -203,6 +203,12 @@ If only one extension has a rule that matches a request, that rule is applied. B
 
 1. If more than one rule matches, the most recently installed extension gets priority.
 
+### Safe rules {: #safe-rules }
```
Comment (sebastianbenz) [inline site/en/docs/extensions/reference/declarativeNetRequest/index.md:206]: What do you think about removing this section? It feels a bit random as it's not clear why we're mentioning it here. How about merging this into the actual limits section and to not use the term "safe rules" at all?

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md [hunk 3]
```diff
@@ -223,7 +229,18 @@ The number of rules available after that depends on how many rules are enabled b
 
 #### Dynamic and session rules {: #dynamic-session-rules }
```
Comment (sebastianbenz) [inline site/en/docs/extensions/reference/declarativeNetRequest/index.md:230]: How about splitting this up into "Static rules", "Dynamic rules" and "Session rules". It's a bit unclear why static rules are standalone, but session and dynamic rule limits are together.

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md (hunk 1 above)
Comment (oliverdunk) [reply site/en/docs/extensions/reference/declarativeNetRequest/index.md:236]: You do :)

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md (hunk 3 above)
Comment (sebastianbenz) [reply site/en/docs/extensions/reference/declarativeNetRequest/index.md:230]: How about describing the pre 121 behavior in a separate callout?

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md (hunk 2 above)
Comment (oliverdunk) [reply site/en/docs/extensions/reference/declarativeNetRequest/index.md]: Done! I did leave the mention of safe rules - since we have the constant called `MAX_NUMBER_OF_UNSAFE_DYNAMIC_RULES`, I think we're locked in to documenting the term safe rule.

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md (hunk 3 above)
Comment (oliverdunk) [reply site/en/docs/extensions/reference/declarativeNetRequest/index.md]: I've moved things around a bit. Let me know what you think. Unfortunately a little tricky with changes in both Chrome 120 and Chrome 121 but tried to find a good compromise.

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md [hunk 4]
```diff
@@ -221,9 +221,23 @@ manifest field to limit which Chrome versions can install your extension.
...
+[`MAX_NUMBER_OF_SESSION_RULES`](#property-MAX_NUMBER_OF_SESSION_RULES).
+
+Before Chrome 120, there was a limit of 5000 combined dynamic and session rules.
+
+#### Dynamic rules {: #dynamic-rules }
+
+An extension can have at least 5000 dynamic rules. This is exposed as the
+[`MAX_NUMBER_OF_UNSAFE_DYNAMIC_RULES`](#property-MAX_NUMBER_OF_UNSAFE_DYNAMIC_RULES).
+
+Starting in Chrome 121, there is a larger limit of 30,000 rules available for safe dynamic rules,
```
Comment (sebastianbenz) [inline site/en/docs/extensions/reference/declarativeNetRequest/index.md:236]: Do these include the 5000 unsafe rules?

**Review by sebastianbenz — APPROVED** [review]

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md (hunk 3 above)
Comment (sebastianbenz) [reply site/en/docs/extensions/reference/declarativeNetRequest/index.md]: Much better - thanks!

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md (hunk 4 above)
Comment (oliverdunk) [reply site/en/docs/extensions/reference/declarativeNetRequest/index.md:236]: It does. I added a note on that, hopefully it helps.

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md [hunk 5]
```diff
@@ -221,9 +221,24 @@ manifest field to limit which Chrome versions can install your extension.
...
+exposed as the [`MAX_NUMBER_OF_DYNAMIC_RULES`](#property-MAX_NUMBER_OF_DYNAMIC_RULES). Safe rules
+are defined as rules with an action of `block`, `allow`, `allowAllRequests` or `upgradeScheme`. Any
+unsafe rules added within the limit of 5000 will also count towards this limit.
+
+Before Chrome 120, there was a limit of 5000 combined dynamic and session rules.
```
Comment (AmySteam) [inline site/en/docs/extensions/reference/declarativeNetRequest/index.md:241]:
```suggestion
Before Chrome 120, there was a 5000 combined dynamic and session rules limit.
```

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md [hunk 6]
```diff
@@ -221,9 +221,24 @@ manifest field to limit which Chrome versions can install your extension.
...
 
-The limits applied to dynamic and session rules are simpler than static rules. The total number of both cannot exceed 5000. This is called the [`MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES`](#property-MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES).
+An extension can have up to 5000 session rules. This is exposed as the
+[`MAX_NUMBER_OF_SESSION_RULES`](#property-MAX_NUMBER_OF_SESSION_RULES).
+
+Before Chrome 120, there was a limit of 5000 combined dynamic and session rules.
+
+#### Dynamic rules {: #dynamic-rules }
+
+An extension can have at least 5000 dynamic rules. This is exposed as the
```
Comment (AmySteam) [inline site/en/docs/extensions/reference/declarativeNetRequest/index.md:233]: The flow of this section is a bit hard to follow. I am writing the facts in bullets to check if I'm understanding correctly:

**Dynamic rule types:**
- **Safe Rules**: those with actions of block, allow, allowAllRequests, or upgradeScheme.
- **Unsafe Rules**: any rules with actions not included under the safe rules.

**Chrome 121 and Later:**
- A higher limit of 30,000 for safe dynamic rules, indicated as `MAX_NUMBER_OF_DYNAMIC_RULES`.
- The limit of 5,000 unsafe dynamic rules, denoted as `MAX_NUMBER_OF_UNSAFE_DYNAMIC_RULES`, also contributes towards the 30,000 safe rule limit.

**Before Chrome 121:**
A combined limit of 5,000 for both dynamic and session rules, without a specific distinction between safe and unsafe rules.

**Review by AmySteam — COMMENTED** [review]
Hey Oliver, I left a clarifying comment to see if we can structure the content in a way that is more explicit.

File: site/en/docs/extensions/reference/declarativeNetRequest/index.md (hunk 6 above)
Comment (oliverdunk) [reply site/en/docs/extensions/reference/declarativeNetRequest/index.md:233]: This looks good to me. Glad you could follow at least, but happy to iterate. Agree this is a lot.

Comment (AmySteam) [reply site/en/docs/extensions/reference/declarativeNetRequest/index.md:233]: Let's leave it as is for the MVP.

**Review by AmySteam — APPROVED** [review]
LGTM 👍
